src/Microsoft.Azure.SignalR/HubHost/ServiceLifetimeManagerBase.cs

```diff
@@ -267,6 +288,86 @@ public override Task RemoveFromGroupAsync(string connectionId, string groupName,
+        {
+            if (IsInvalidArgument(connectionId))
+            {
+                throw new ArgumentException(NullOrEmptyStringErrorMessage, nameof(connectionId));
+            }
+
+            if (IsInvalidArgument(methodName))
+            {
+                throw new ArgumentException(NullOrEmptyStringErrorMessage, nameof(methodName));
+            }
+
+            var invocationId = ClientInvocationManager.Caller.GenerateInvocationId(connectionId);
+
+            try
+            {
+                var message = AppendMessageTracingId(new ClientInvocationMessage(invocationId, connectionId, _callerId, SerializeAllProtocols(methodName, args, invocationId)));
+                await WriteAsync(message);
+                if (ClientConnectionManager.ClientConnections.TryGetValue(connectionId, out var clientConnectionContext))
+                {
+                    var instanceId = clientConnectionContext.InstanceId;
+                    var task = ClientInvocationManager.Caller.AddInvocation<T>(connectionId, invocationId, instanceId, cancellationToken);
+                    return await task;
+                }
+                else
+                {
+                    throw new ArgumentException(NullOrEmptyStringErrorMessage, nameof(connectionId));
+                }
+            }
+            catch (Exception)
+            {
+                throw;
+            }
+        }
+
+        // Only route server will reach here
+        public override async Task SetConnectionResultAsync(string connectionId, CompletionMessage result)
+        {
+            if (IsInvalidArgument(connectionId))
+            {
+                throw new ArgumentException(NullOrEmptyStringErrorMessage, nameof(connectionId));
+            }
+            if (ClientConnectionManager.ClientConnections.TryGetValue(connectionId, out var clientConnectionContext))
+            {
+                // Determine which manager (Caller / Router) the `result` belongs to.
+                // `TryCompletionResult` returns false when the corresponding invocation is not existing.
+                IClientResultsManager clientResultsManager = null;
+                if (ClientInvocationManager.Caller.TryCompleteResult(connectionId, result))
+                {
+                    clientResultsManager = ClientInvocationManager.Caller;
+                }
+                if (ClientInvocationManager.Router.TryCompleteResult(connectionId, result))
+                {
+                    clientResultsManager = ClientInvocationManager.Router;
+                }
+
+                // Block unknown `results` which belongs to neither Caller nor Router
+                if (clientResultsManager != null)
+                {
+                    var protocol = clientConnectionContext.Protocol;
+                    var message = AppendMessageTracingId(new ClientCompletionMessage(result.InvocationId, connectionId, _callerId, protocol, SerializeCompletionMessage(result, protocol)));
```

Comment:
What is ClientCompletionMessage for when it is from Caller? There is an additional result seralization, is it neccessory when Caller case?

Reply:
1. Caller need to send back `ClientCompletionMessage` to service as well.  Because service will rely on `ClientCompeletionMessage` to cleanup. 
See  [here](https://msazure.visualstudio.com/One/_git/OSSServices-SignalR-Service?path=/src/Microsoft.Azure.SignalR.ServiceRuntime/LifetimeManager/ClientConnectionLifetimeManager.cs&version=GBdev&line=802&lineEnd=803&lineStartColumn=1&lineEndColumn=1&lineStyle=plain&_a=contents)
2. Thus this seralization is also necessary when caller case.

Reply:
Probably we could simply the case for Caller: make payload empty instead of calling `SerializeCompletionMessage`?